From 19adb7b4de01049d0307da079fb5f31ca59b1a91 Mon Sep 17 00:00:00 2001
From: Archer <archer@beezig.eu>
Date: Thu, 13 Jan 2022 01:49:56 +0100
Subject: [PATCH] event: Add PreSignChangeEvent

This commit adds an event that's called *before* an incoming sign change
is processed by the server. This event is useful to handle invalid sign
changes that are dismissed by the server before the regular
SignChangeEvent is called.

diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 2c6c1e2bc..1aa0d226c 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -4,7 +4,7 @@ import com.google.common.collect.Lists;
 import com.google.common.primitives.Doubles;
 import com.google.common.primitives.Floats;
 import dev.rocco.kig.paper.api.event.DisconnectReason;
-import dev.rocco.kig.paper.api.event.PlayerInteractUpdateEvent;
+import dev.rocco.kig.paper.api.event.PreSignChangeEvent;
 import dev.rocco.kig.paper.impl.event.KigEvents;
 import io.netty.buffer.Unpooled;
 import io.netty.util.concurrent.Future;
@@ -1939,6 +1939,22 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
     public void a(PacketPlayInUpdateSign packetplayinupdatesign) {
         if (this.player.dead) return; // CraftBukkit
         PlayerConnectionUtils.ensureMainThread(packetplayinupdatesign, this, this.player.u());
+        // KigPaper start
+        Player player = this.server.getPlayer(this.player);
+        int x = packetplayinupdatesign.a().getX();
+        int y = packetplayinupdatesign.a().getY();
+        int z = packetplayinupdatesign.a().getZ();
+        IChatBaseComponent[] aichatbasecomponent = packetplayinupdatesign.b();
+        String[] lines = new String[4];
+
+        for (int i = 0; i < aichatbasecomponent.length; ++i) {
+            lines[i] = EnumChatFormat.a(aichatbasecomponent[i].c());
+        }
+        PreSignChangeEvent preSignChangeEvent =
+                new PreSignChangeEvent(new Location(player.getWorld(), x, y, z), player, lines);
+        this.server.getPluginManager().callEvent(preSignChangeEvent);
+        if (preSignChangeEvent.isCancelled()) return;
+        // KigPaper end
         this.player.resetIdleTimer();
         WorldServer worldserver = this.minecraftServer.getWorldServer(this.player.dimension);
         BlockPosition blockposition = packetplayinupdatesign.a();
@@ -1958,18 +1974,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
                 return;
             }
 
-            IChatBaseComponent[] aichatbasecomponent = packetplayinupdatesign.b();
-
             // CraftBukkit start
-            Player player = this.server.getPlayer(this.player);
-            int x = packetplayinupdatesign.a().getX();
-            int y = packetplayinupdatesign.a().getY();
-            int z = packetplayinupdatesign.a().getZ();
-            String[] lines = new String[4];
-
-            for (int i = 0; i < aichatbasecomponent.length; ++i) {
-                lines[i] = EnumChatFormat.a(aichatbasecomponent[i].c());
-            }
             SignChangeEvent event = new SignChangeEvent((org.bukkit.craftbukkit.block.CraftBlock) player.getWorld().getBlockAt(x, y, z), this.server.getPlayer(this.player), lines);
             this.server.getPluginManager().callEvent(event);
 
-- 
2.34.1

